### PR TITLE
fix wrong email when use gitea as OAuth2 provider

### PR DESCRIPTION
**Pull request by renothing**
when you use gitea as OAuth2 provider, the /api/v1/user should return
user primary email as identifier, which is unique in OAuth2 clients.

### CODE REVIEW

**Comment by lunny** [comment]
No, it's for security reason

**Comment by renothing** [comment]
> No, it's for security reason

I think this api call only request by user itself. isn't it?

**Comment by lunny** [comment]
@renothing It will be requested by a third-party application of gitea instance. If user set his email private, we should not send the email to third-party application which could save it and you will never know if it happened.

**Comment by renothing** [comment]
@lunny I think `private` is just for gitea users internal, especially from frontend. when you use gitea as a OAuth2 provider, the username and email should be part of basic data. the others most popular like github/gitlab also provider email too.  current there is no scope control in gitea,  when the user grant authentication to 3rd, the 3rd application even can update email of the user.

I checked the code and I know what you said about security, it just because `/api/v1/users/` use the same method.

I think we should identify whether it is yourself. or use new endpoint for this, like `/api/v1/me` for OAuth2 consumer.

BTW, I think return return nothing better rather than a fake mail.

**Comment by lunny** [comment]
@renothing Right, I think if we can know it's the login user himself, he can get himself email.

**Comment by renothing** [comment]
@lunny  already update. please review it.

File: routers/api/v1/convert/convert.go [hunk 1]
```diff
@@ -224,12 +224,15 @@ func ToUser(user *models.User, signed, admin bool) *api.User {
...
-		AvatarURL: user.AvatarLink(),
 		FullName:  markup.Sanitize(user.FullName),
+		AvatarURL: user.AvatarLink(),
+		Language:  user.Language,
 		IsAdmin:   user.IsAdmin,
 	}
 	if signed && (!user.KeepEmailPrivate || admin) {
 		result.Email = user.Email
+	} else {
+		result.Email = ""
```
Comment (zeripath) [inline routers/api/v1/convert/convert.go]: result.Email will be "" by default. So this is unnecessary. Go defaults unset strings to "".

**Review by zeripath — CHANGES_REQUESTED** [review]

**Comment by zeripath** [comment]
This is not your fault but I we need to rationalise `user.APIFormat()` and `convert.ToUser()` - they look like they're basically duplicating each other and likely we should be abandoning `user.APIFormat()` and making `convert.ToUser()` use the hidden email address as appropriate.

Further, We should be very careful about what we decide to send to non-logged in users and so on:

* I don't think leaking whether a user is an Admin to non-logged in users is reasonable or good practice.
* Language is probably ok
* LastLoginTime and CreatedTime should not be sent to non-logged in users.

File: routers/api/v1/convert/convert.go (hunk 1 above)
Comment (renothing) [reply routers/api/v1/convert/convert.go]: > result.Email will be "" by default. So this is unnecessary. Go defaults unset strings to "".

can you tell me why return fake mail by default, it's very confused for end user.

**Comment by renothing** [comment]
> This is not your fault but I think we need to rationalise `user.APIFormat()` and `convert.ToUser()` - they look like they're basically duplicating each other and likely we should be abandoning `user.APIFormat()` and making `convert.ToUser()` use the hidden email address as appropriate.
> 
> Further, We should be very careful about what we decide to send to non-logged in users and so on:
> 
> * I don't think leaking whether a user is an Admin to non-logged in users is reasonable or good practice.
> * Language is probably ok
> * LastLoginTime and CreatedTime should not be sent to non-logged in users.

thanks for your reply. I revert convert.ToUser just keep as default . I'm not familiar with golang. this is my first pr.  and I see `LastLogin` and `Created` added in `convert.ToUser` within master branch.

**Comment by zeripath** [comment]
As I say it's not your fault - I've also touched this code - but I think we do need to do something. Your changes were not wrong. The background is wrong.

**Comment by lunny** [comment]
@renothing For a package dependency optimization, I think we can remove `user.APIFormat()`and always use `convert.ToUser` instead.

**Comment by renothing** [comment]
> @renothing For a package dependency optimization, I think we can remove `user.APIFormat()`and always use `convert.ToUser` instead.

I think we should still keep APIFormat in current. look at this scenes. if the caller make a webhook request, the primary email will be leak away if you use `convert.ToUser`. There are still many requests that do not require email at all like `ListSubscribers`

and I saw Too many `APIFormat` call in total code. conservatively it's better to keep pr smaller.

**Comment by zeripath** [comment]
I've just seen that this based on the 1.8 branch which is in maintenance only. This would need to be based on master.

**Comment by zeripath** [comment]
OK So I've had a look at removing user.APIFormat or changing it to be like convert.ToUser, and you're right it's a total nightmare - this needs some more thought.

**Pull request closed without merging**